lib/spack/spack/util/spack_yaml.py

```diff
@@ -60,10 +65,56 @@ class syaml_str(str):
     __repr__ = str.__repr__
 
 
+class syaml_int(int):
+    __repr__ = str.__repr__
+
+
+#: primitive type -> syaml type mapping
+syaml_types = {
+    syaml_str: string_types,
+    syaml_int: int,
+    syaml_dict: dict,
+    syaml_list: list,
+}
+
+
+def syaml_type(obj):
+    """Get the corresponding syaml wrapper type for a primitive type.
+
+    Return:
+        (object): syaml-typed copy of object, or the obj if no wrapper
+    """
+    for syaml_t, t in syaml_types.items():
+        if type(obj) is not bool and isinstance(obj, t):
+            return syaml_t(obj) if type(obj) != syaml_t else obj
+    return obj
+
+
+def markable(obj):
+    """Whether an object can be marked."""
+    return type(obj) in syaml_types
+
+
 def mark(obj, node):
     """Add start and end markers to an object."""
-    obj._start_mark = node.start_mark
-    obj._end_mark = node.end_mark
+    if not markable(obj):
+        return
+
+    if hasattr(node, 'start_mark'):
+        obj._start_mark = node.start_mark
+    elif hasattr(node, '_start_mark'):
+        obj._start_mark = node._start_mark
```

Comment:
Side note on style (not directly introduced by this PR). We are doing a lot of reads and writes of two "private" attributes:
```python
obj._start_mark
obj._end_mark
``` 
Probably it would be nicer to make the attribute names not start with an underscore.

Reply:
This was done to avoid conflicts with YAML's own markings.  I suppose I could've picked another prefix, but it starts with `_` to avoid conflicts with vanilla YAML, not specifically to be private.  I suppose we could call it something more specific/obvious but I think we can update that later, as it's a bigger and mostly cosmetic change.